### DISCUSSION TITLE
Create and Create Another, with a form refresh in between

### DISCUSSION THREAD

**Discussion by intrepidws** [filamentphp/filament#1756] [Help] [answered]
I have a Create form with a text input that has a dynamically populated datalist of available options. I'd like the user to be able to use the "Create & Create Another" functionality but I need the form to refresh after the initial create, so that the datalist observes any changes based on the initial create.

Is this possible?

**Comment by danharrin** [comment]
Can I see the code you're using for the dependant datalist options?

**Reply by intrepidws** [reply]
Absolutely.

```php
public static function form(Form $form): Form
{
    return $form
        ->schema([
            Forms\Components\TextInput::make('location')
                ->maxLength(255)
                ->datalist(function (Livewire $livewire) {
                    return array_merge(
                        ['Front Yard', 'Back Yard', 'Side Yard', 'Driveway'],
                        $livewire->ownerRecord->transformers->pluck('pivot.location')->toArray(),
                        $livewire->ownerRecord->fixtures->pluck('pivot.location')->toArray(),
                    );
                }),
            Forms\Components\Select::make('quantity')
                ->options([1 => 1, 2 => 2, 3 => 3, 4 => 4, 5 => 5, 6 => 6, 7 => 7, 8 => 8, 9 => 9, 10 => 10])
                ->required(),
            Forms\Components\TextInput::make('fixture')
                ->maxLength(255),
            Forms\Components\TextInput::make('lamp')
                ->maxLength(255),
            Forms\Components\TextInput::make('description')
                ->maxLength(255),
        ]);
}
```

**Reply by intrepidws** [reply]
I tried adding `$livewire->ownerRecord->fresh(['fixtures', 'transformers']);` in there but no dice.

**Reply by danharrin** [reply]
So is the datalist closure executing at the correct times, the problem is with the information being out of date from the ownerRecord? Or is the datalist just not reloading options

**Reply by intrepidws** [reply]
I did some testing and can confirm that the code within the `datalist()` method actually runs again, after I click "Create & create another" but the options in the datalist don't actually update.

**Reply by danharrin** [reply]
Can you track down exactly which model/s or relationship/s are out of date there?

**Reply by intrepidws** [reply]
It looks like the main `$livewire->ownerRecord` model as well as the relevant relationship models are all out of date. For instance, if I output a relationship count right inside the datalist function, here is the output at various stages:

1. After clicking "Create" button the first time to launch the modal: 0
2. After filling out the modal form and clicking the "Create & create another" button: 0
3. After clicking the "Cancel" button from within the modal and then clicking the "Create" button again to re-launch the modal: 1

I was able to get the relationship count to update at step 2 by forcing an update of the model at the beginning of the datalist function. by using `$livewire->ownerRecord->refresh();`. Maybe this should happen automatically after clicking the "Create & create another" button?

**Reply by intrepidws** [reply]
Sorry it occurs to me that I might not have specified that this is part of a `BelongsToManyRelationManager` not a regular Resource.

**Reply by danharrin** [reply]
I'm glad you were able to sort it with `$livewire->ownerRecord->refresh();`. Not sure if this should be default behaviour, since it will result in another DB query that most users won't need?

**Comment by intrepidws** [comment, marked as answer]
Just to close the loop here, I was able to use `$livewire->ownerRecord->refresh();` to fix my issue. Code sample below:

```php
public static function form(Form $form): Form
{
    return $form
        ->schema([
            Forms\Components\TextInput::make('location')
                ->maxLength(255)
                ->datalist(function (Livewire $livewire) {
                    $livewire->ownerRecord->refresh();

                    return array_merge(
                        ['Front Yard', 'Back Yard', 'Side Yard', 'Driveway'],
                        $livewire->ownerRecord->fixtures->pluck('pivot.location')->toArray(),
                    );
                }),
        ]);
}
```
